### ISSUE TITLE
[no-unnecessary-condition] False error on uninitialized variable

### ISSUE THREAD

**Issue by Retro64** [typescript-eslint/typescript-eslint#4513]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

[Playground](https://typescript-eslint.io/play#ts=5.3.2&fileType=.tsx&code=DYUwLgBAZg9jBcEDOYBOBLAdgcwNwCh8BjGTFZcAMTggF4IAKASkQDcZ0ATOgPggG98ECOiiNYMAPwA6UDjAALCAB4IARiYChw6DXoByCfoLCAvvlMEgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

```JSON
{
  "rules": {
      "@typescript-eslint/no-unnecessary-condition": [
      "error"
    ],
  }
}
```

```TS
let foo: string;

const setFoo = (): void => {
  if (foo?.length < 1) {
    foo = 'foo';
  }
};

```

tsconfig
```
{
  "compilerOptions": {
    "outDir": "./lib",
    "module": "commonjs",
    "target": "es2015",
    "lib": ["es6", "es7"],
    "sourceMap": true,
    "allowJs": false,
    "moduleResolution": "node",
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strict": true,
    "strictNullChecks": true,
    "suppressImplicitAnyIndexErrors": true,
    "noUnusedLocals": true,
    "noErrorTruncation": true,
    "pretty": true,
    "declaration": true
  },
  "include": [
    "src/**/*"
  ]
}
```

**Expected Result**

Existence check on a possibly uninitialized variable should not trigger no-unnecessary-condition

For sure, the type might be extended to `let foo = string | undefined;` as workaround, but that actually looks a bit uncommon for a let definition, where not initializing might be intended.

**Actual Result**
`Unnecessary optional chain on a non-nullish value  @typescript-eslint/no-unnecessary-condition`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.0` |
| `@typescript-eslint/parser`        | `5.10.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.7.0` |
| `ts-node`                             | `10.4.0` |

**Comment by bradzacher** [comment]
The rule was written before optional chaining, and this "use before assigned" case was never added.

It's very, very rare for people to write "use before assigned" variables, and even rarer that people use optional chaining with them.

**Comment by sosoba** [comment]
Another example of this same bug: 
```ts
let a = false;
const b = () => a = true;
b();
if (a) { // Unnecessary conditional, value is always falsy.
}
```

**Comment by bradzacher** [comment]
@sosoba - no, that is not the same problem.
There are many existing duplicate issues about your issue:
https://github.com/typescript-eslint/typescript-eslint/issues/4624 https://github.com/typescript-eslint/typescript-eslint/issues/4501 https://github.com/typescript-eslint/typescript-eslint/issues/4315 https://github.com/typescript-eslint/typescript-eslint/issues/3957 https://github.com/typescript-eslint/typescript-eslint/issues/3732

**Comment by armano2** [comment]
this has to be fixed in typescript, uninicialized variable should be reported by typechecker as `X | undefined` not `X` or we should be able to receive information that variable is uninitialized in some way

https://www.typescriptlang.org/play?ssl=8&ssc=1&pln=1&pc=1#code/DYUwLgBGIM5gXBOAnAlgOwOYG4BQuAzAV3QGMxUB7dCADwgAoBKCAb1wk4mXCORuhwAdKCxgAFngC++WszxA

```ts
let test: string;
function x () {
    return test.length;
}
x();
```

this code is going to compile without error but, its actually going to produce runtime error

**Comment by armano2** [comment]
that's good idea to use scopes for that

there is only one issue with external variables, eg. exported from another file, but that should be even rarer case

**Comment by bradzacher** [comment]
Across module boundaries there be Dragons.
So I'm more than happy for us to fail the case!

**Comment by JoshuaKGoldberg** [comment]
Note from #9206 that `var` is also a relevant case:

```ts
if (Math.random()) {
  var x = 1;
}
x!;
export {}
```

**Comment by kirkwaiblinger** [comment]
I think the rule's behavior is correct in this case. And I think that `let x: T;` should _not_ be treated as `T | undefined`.

IMO if you plan to _read_ a variable in its uninitialized state, you should use `let x: T | undefined;` (or, more verbosely, `let x: T | undefined = undefined;`). If your intention is that it will always be explicitly set _before_ being read (whether or not this is provable from static analysis, e.g. due to setting the variable in a closure), `let x: T` is appropriate.

**Comment by kirkwaiblinger** [comment]
Note also that the code which is proposed to be allowed in this issue _hides a type error from TS that would be reported if the rule were followed_. The only reason that the code in the issue report is allowed by TS at all, is that TS assumes that `foo` is _not_ `undefined`. If it were possibly undefined, `foo?.length` would have type `string | undefined`, which is an error to compare to a number with the `<` operator. ([playground](https://typescript-eslint.io/play/#ts=5.3.2&fileType=.tsx&code=DYUwLgBAZg9jBcEDOYBOBLAdgcwgHwgFdMATEKLEEgbgChaBjGTFZcAMTggF4IAKAJSIAbjHQkeAPggBvWhAjoo-WDAD8AOlA4wACwgAeCAEYBs%2BQuhdeAclU26CgL60n1IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))
```ts
let foo: string | undefined;

const setFoo = (): void => {
  if (foo?.length < 1) {
//    ~~~~~~~~~~~~~~~ TS ERROR 18048: 'foo.length' is possibly 'undefined'. 4:7 - 4:18     
    foo = 'foo';
  }
};
```

Furthermore, if we remove the optional chain, it's _not_ a type error ([playground](https://typescript-eslint.io/play/#ts=5.3.2&fileType=.tsx&code=DYUwLgBAZg9jBcEDOYBOBLAdgcwNwCh8BjGTFZcAMTggF4IAKASkQDcZ0ATOgPggG98ECOiiNYMAHSgcYABYQAPBACMTAUOHQa9AOQTdBYQF98x3EA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))
```ts
let foo: string;

const setFoo = (): void => {
  if (foo.length < 1) {
    foo = 'foo';
  }
};
```
Which _cannot_ be the case for removing an unnecessary optional chain on an accurately typed variable.

This is all to say, `let foo: string;` _means_ "by the time you read `foo` it will be a `string`", and TS makes correctness assumptions about that with much broader implications than just the behavior in this rule. Therefore, regardless of how one feels about that decision on TS's part, I think it would be wrong to retcon it to mean anything else for this rule. Use `let foo: string | undefined;`.

**Comment by kirkwaiblinger** [comment]
> Note from #9206 that `var` is also a relevant case:
> 
> ```ts
> if (Math.random()) {
>   var x = 1;
> }
> if (x) {}
> ```

Fortunately, that's a TS use-before-define error 🙂, so I think we can avoid the mental strain of thinking about `var`, unless we can think of another case where it matters.

**Comment by JacobLey** [comment]
I've run into a related issue, one that I _think_ abides by strict TS usage.

```
class ConsistentRand {
    #rand?: number;

    getCachedRand() {
        // Error reported here
        this.#rand ??= Math.random();
        return this.#rand;
    }
}
```

Typescript acknowledges that `this.#rand` may be undefined (doesn't expect initialization) but it is still reported as 
> Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined

Re-writing as 
```
this.#rand = this.#rand ?? Math.random();
```

Removes this error, despite it logically being the exact same. In fact, this will fail on unrelated [logical-assignment-operators](https://eslint.org/docs/latest/rules/logical-assignment-operators).

**Comment by bradzacher** [comment]
@JacobLey - that [does not reproduce on the playground](https://typescript-eslint.io/play/#ts=5.4.5&showAST=es&fileType=.tsx&code=MYGwhgzhAEDCD2A7CBLCAXApo9AlMiAJtAN4BQ00AxAE4GED8AXNIgK4C2ARpjQNxkK0AOaZ0sMMAAWmQviIAKAJSkhlAPTroAURo14NaDUwAHA1mIzja6OiloAdLXrQGDALzQAsmDsO6RPAcygKUlMbobDSItvYQTgGEodAAvmRpQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

any such issues are unrelated to this issue and should have been fixed as part of #6635.
please file a new issue with a repro and we can investigate.

**Comment by JacobLey** [comment]
[Small adjustments to playground](https://typescript-eslint.io/play/#ts=5.5.3&fileType=.tsx&code=MYGwhgzhAEDCD2A7CBLCAXApo9AlMiAJtAN4BQ00AxAE4GED8AXNIgK4C2ARpjQNxkK0AOaZ0sMMAAWmQviIAKAJSkhldFLQA6WvWgMGAXmgBZMBq10i8DsoGVKNMWxqJoG7bqL3oAXzL%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2ToAPAIbNcBYkNQAFTLERY8ATwAqi%2BWUod0vAL4gdQA&tokens=false) and I am able to reproduce.

I believe `exactOptionalPropertyTypes=true` is the secret.

Will open a separate issue so it can receive a separate investigation.

**Comment by kirkwaiblinger** [comment]
> I think the rule's behavior is correct in this case. And I think that `let x: T;` should _not_ be treated as `T | undefined` for the purposes of this rule, because it is not treated that way by TS.

Since I wrote this, it was pointed out to me by @bradzacher that this rule autofixes this case, which can cause unsafe behavior. That's a very good point!

I still believe strongly that the _report_ is correct, but I agree that it's problematic to autofix this case. I believe the appropriate resolution would be to detect this case and downgrade the autofix to a suggestion. 

I'd like to open this issue to PRs for that resolution if there are no objections?

**Comment by kirkwaiblinger** [comment]
filed https://github.com/typescript-eslint/typescript-eslint/issues/9565 to propose flagging the code in the issue report for failing to initialize `foo`
